keps/sig-network/2594-multiple-cluster-cidrs/README.md

```diff
@@ -0,0 +1,882 @@
+    -   Can not modify a `NodeSelector` to deselect a node that is using that
+        range.
+
+#### Example Config
+
+```
+[
+{
+    # For existing clusters this is the same as ClusterCIDR
+    CIDR:            "10.0.0.0/8",
+    # Default for nodes not matching any other rule
+    NodeSelector:    nil,
+    # For existing API this is the same as NodeCIDRMaskSize
+    PerNodeMaskSize: 24,
+},
+{
+    CIDR:            "172.16.0.0/14",
+    # Another range, also allocate-able to any node
+    NodeSelector:    nil,
+    PerNodeMaskSize: 24,
+},
+{
+    CIDR:            "10.0.0.0/8",
+    NodeSelector:    { key: "np" op: "IN" value:["np1"] },
+    PerNodeMaskSize: 26,
+},
+{
+    CIDR:            "192.168.0.0/16",
+    NodeSelector:    { key: "np" op: "IN" value:["np2"] },
+    PerNodeMaskSize: 26,
+},
+{
+    CIDR:            "5.2.0.0/16",
+    NodeSelector:    { "np": "np3" },
+    PerNodeMaskSize: 20,
+},
+{
+    CIDR:            "fd12:3456:789a:1::/64",
+    NodeSelector:    { "np": "np3" },
+    PerNodeMaskSize: 112,
+},
+...
+]
+```
+
+Given the above config, a valid potential configuration might be:
+
+```
+{“np”: “np1”} --> "10.0.0.0/26"
+{“np”: “np2”} --> "192.16.0.0/26"
+{“np”: “np3”} --> "5.2.0.0/20", "fd12:3456:789a:1::/112"
+{“np”: “np4”} --> "172.16.0.0/24"
+```
+
+### Controller
+
+Implement a new
+[NodeIPAM controller](https://github.com/kubernetes/kubernetes/tree/master/pkg/controller/nodeipam)
+The controller will set up watchers on the `ClusterCIDRConfig` objects and the
+`Node` objects.
+
+#### Data Structures
+
+We will use maps to store the allocated ranges and which node is using the
+range. Because the number of nodes is expected to be on the order of thousands,
+more sophisticated data structure are likely not required.
+
+Prior investgations [here](https://github.com/kubernetes/kubernetes/pull/90184)
+suggest that maps storing allocations will perform well under the number of
+nodes we expect.
+
+#### Startup
+
+-   Fetch list of `ClusterCIDRConfig` and build internal data structure
+-   If node `ClusterCIDRConfig` are specified, fetch the values of the
+    cluster-cidr and node-cidr-mask-size arguments passed to
+    kube-controller-manager.
+-   Fetch list of `Node`s. Check each node for `PodCIDRs`
+    -   If `PodCIDR` is set, mark the allocation in the internal data structure
+        and store this association with the node.
+    -   If `PodCIDR` is set, but is not part of one of the tracked
+        `ClusterCIDRConfig`s, emit a K8s event but do nothing.
+    -   If `PodCIDR` is not set, save Node for allocation in the next step.
+        After processing all nodes, allocate ranges to any nodes without Pod
+        CIDR(s) [Same logic as Node Added event]
+
+#### Reconciliation Loop
+
+This go-routine will watch for cleanup operations and failed allocations and
+continue to try them in the background.
+
+For example if a Node can't be allocated a PodCIDR, it will be periodically
```

Comment:
so, this is for handling a case where we don't have free cidrs, and then a node is deleted so we suddenly have one free cidr that we can use for this node that was waiting ... it is like an infinite requeue thing for the controller , right?

Reply:
deleting, i just read the next section :)